app/channel.go

```diff
@@ -45,71 +45,54 @@ func (a *App) JoinDefaultChannels(teamId string, user *model.User, shouldBeAdmin
-		}
-
-		if *a.Config().ServiceSettings.ExperimentalEnableDefaultChannelLeaveJoinMessages {
-			if requestor == nil {
-				if err := a.postJoinTeamMessage(user, townSquare); err != nil {
-					mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
-				}
-			} else {
-				if err := a.postAddToTeamMessage(requestor, user, townSquare, ""); err != nil {
-					mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
-				}
-			}
-		}
+	defaultChannelList := []string{"town-square"}
 
-		a.InvalidateCacheForChannelMembers(result.Data.(*model.Channel).Id)
+	if len(a.Config().TeamSettings.ExperimentalDefaultChannels) == 0 {
+		defaultChannelList = append(defaultChannelList, "off-topic")
+	} else {
+		defaultChannelList = append(defaultChannelList, a.Config().TeamSettings.ExperimentalDefaultChannels...)
 	}
 
-	if result := <-a.Srv.Store.Channel().GetByName(teamId, "off-topic", true); result.Err != nil {
-		err = result.Err
-	} else if offTopic := result.Data.(*model.Channel); offTopic.Type == model.CHANNEL_OPEN {
+	for _, channelName := range defaultChannelList {
+		if result := <-a.Srv.Store.Channel().GetByName(teamId, channelName, true); result.Err != nil {
+			err = result.Err
+		} else {
 
-		cm := &model.ChannelMember{
-			ChannelId:   offTopic.Id,
-			UserId:      user.Id,
-			SchemeUser:  true,
-			SchemeAdmin: shouldBeAdmin,
-			NotifyProps: model.GetDefaultChannelNotifyProps(),
-		}
+			channel := result.Data.(*model.Channel)
 
-		if cmResult := <-a.Srv.Store.Channel().SaveMember(cm); cmResult.Err != nil {
-			err = cmResult.Err
-		}
-		if result := <-a.Srv.Store.ChannelMemberHistory().LogJoinEvent(user.Id, offTopic.Id, model.GetMillis()); result.Err != nil {
-			mlog.Warn(fmt.Sprintf("Failed to update ChannelMemberHistory table %v", result.Err))
-		}
+			if channel.Type != model.CHANNEL_OPEN {
+				continue
+			}
 
-		if requestor == nil {
-			if err := a.postJoinChannelMessage(user, offTopic); err != nil {
-				mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
+			cm := &model.ChannelMember{
+				ChannelId:   channel.Id,
+				UserId:      user.Id,
+				SchemeUser:  true,
+				SchemeAdmin: shouldBeAdmin,
+				NotifyProps: model.GetDefaultChannelNotifyProps(),
 			}
-		} else {
-			if err := a.PostAddToChannelMessage(requestor, user, offTopic, ""); err != nil {
-				mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
+
+			if cmResult := <-a.Srv.Store.Channel().SaveMember(cm); cmResult.Err != nil {
+				err = cmResult.Err
+			}
+			if result := <-a.Srv.Store.ChannelMemberHistory().LogJoinEvent(user.Id, channel.Id, model.GetMillis()); result.Err != nil {
+				mlog.Warn(fmt.Sprintf("Failed to update ChannelMemberHistory table %v", result.Err))
+			}
+
+			if *a.Config().ServiceSettings.ExperimentalEnableDefaultChannelLeaveJoinMessages {
+				if requestor == nil {
+					if err := a.postJoinTeamMessage(user, channel); err != nil {
```

Comment:
@jasonblais This will post `User has joined the team` in all default channels while currently we just post that in Town Square (Off Topic just has the usual "has joined the channel" message). Should we re-add the old behaviour where only town square gets the "joined the team" message?

We could just check if `channelName == model.DEFAULT_CHANNEL` to determine whether or not to call `a.postJoinTeamMessage` vs `a.postJoinChannelMessage`

Reply:
@hmhealey Yes, having `User has joined the team` in Town Square only is preferred. Good catch.